[Audio] Tour of applications (#339)
huggingface/course: chapters/en/chapter10/5.mdx

[//]: # On review, this sub-section might be too involved. Happy to omit it and dive straight into how to load a dataset.

### How to load an audio dataset
Before we can handle any audio datasets, we need to make sure we have the right dependencies installed. Specifically, we'll need the "Audio" feature from datasets: https://huggingface.co/docs/datasets/installation#audio. This should take care of the main Python packages we need to read audio files from byte form and convert them to arrays.

Right! Now we're ready to go ahead and download our data. For this tutorial, we'll use the [_LibriSpeech ASR_](https://huggingface.co/datasets/librispeech_asr) corpus. LibriSpeech is one of the most popular datasets for benchmarking speech recognition systems in both academia and industry. It consists of approximately 1000 hours of narrated audiobooks collected from the (LibriVox)[https://librivox.org/] project.

> Let’s take a look at the all-time leaderboard on Papers with Code to get a feel for numbers: https://paperswithcode.com/sota/speech-recognition-on-librispeech-test-clean

Reviewer: ```suggestion
Let’s take a look at the [all-time leaderboard](https://paperswithcode.com/sota/speech-recognition-on-librispeech-test-clean) on Papers with Code to get a feel for current model performances.
```

 'text': 'CHAPTER SIXTEEN I MIGHT HAVE TOLD YOU OF THE BEGINNING OF THIS LIAISON IN A FEW LINES BUT I WANTED YOU TO SEE EVERY STEP BY WHICH WE CAME I TO AGREE TO WHATEVER MARGUERITE WISHED',
 'speaker_id': 374,
 'chapter_id': 180298,
 'id': '374-180298-0000'}
```

Alright! We can see that we have our target text sample ready - it looks very much like it's taken from an audiobook! But what on earth is going on with the audio? 🤔

As it turns out, we represent audio data digitally as a 1-dimensional array. Each array value denotes the amplitude of our audio signal at a particular time step. From the amplitude information, we can reconstruct the frequency spectrum of the audio and recover all the acoustic features.

Since our audio input is continuous, we can't represent an amplitude value for every possible time step. Instead, we _sample_ amplitude values at fixed time steps.

The interval with which we sample our audio is known as the _sampling rate_. For our dataset, we can see that the sampling rate is 16000, meaning 16000 amplitude values are provided each second. Keep the sampling rate in the back of your mind, it'll be important when we come to processing our data later on!

> Whilst 1-dimensional arrays are a suitable format for machines, they're not much use to us as humans! We want to be able to **listen** to our audio to get a feel for the speech and recording conditions. Here, we have two options:

Reviewer: ```suggestion
While 1-dimensional arrays are a suitable format for machines, they're not of much use to us as humans! We want to be able to **listen** to our audio to get a feel for the speech and recording conditions. Here, we have two options:
```

1. Convert our 1-dimensional array into a human-friendly format (mp3 or wav).
2. Look to the Hugging Face Hub!

> Option 2 seems much easier to me! Let's head over to the LibriSpeech ASR dataset card on the Hugging Face Hub: https://huggingface.co/datasets/librispeech_asr

Reviewer: ```suggestion
Option 2 seems much easier to me! Let's head over to the [LibriSpeech ASR dataset card](https://huggingface.co/datasets/librispeech_asr) on the Hugging Face Hub.
```

> Slap bang in the middle of the dataset card we have exactly what we're looking for: the dataset viewer! The dataset viewer shows us the first 100 samples of any dataset. What's more, it's loaded up the audio samples ready for us to listen to in real-time! If we hit the play button on the first sample, we can listen to the audio and see the corresponding text. Have a scroll through the samples for the train and test sets to get a better feel for the audio data that we're dealing with. You’ll notice how clear the audio is and how well spoken the sentences are. This provides us with a clue as to why the top models do so well on LibriSpeech! The audio conditions are very conducive to high system performance.

Reviewer: ```suggestion
Slap bang in the middle of the dataset card we have exactly what we're looking for: the dataset viewer! The dataset viewer shows us the first 100 samples of any dataset. What's more, it's already loaded the audio samples for us to listen to in real-time! If we hit the play button on the first sample, we can listen to the audio and see the corresponding text. Have a scroll through the samples in the train and test sets to get a better feel for the audio data that we're dealing with. You’ll notice how clear the audio is and how well-spoken the sentences are. This gives us a clue as to why the top models do so well on LibriSpeech! The audio conditions are very conducive to high system performance.
```

### Models for speech recognition

We can decompose speech recognition models into two parts:
1. Encoder: an acoustic model that maps the raw audio input into a sequence of hidden-states.

diagram

2. Decoder: maps the sequence of hidden-states to logits over the vocabulary.

diagram

> The encoder is typically of Wav2Vec2 architecture, introduced in Section 4 of this Chapter. We have some flexibility in our choice of decoder. We could either use a simple linear layer that maps the Wav2Vec2 hidden states directly to output logits over our vocabulary. Or, we could pair our Wav2Vec2 encoder with a decoder model, giving a speech encoder-decoder style model, analogous to the NLP encoder-decoder model introduced in Chapter 1.

Reviewer: ```suggestion
The encoder is typically a Wav2Vec2 architecture, introduced in Section 4 of this Chapter. We have some flexibility in our choice of the decoder. We could use a simple linear layer that maps the Wav2Vec2 hidden states directly to output logits over our vocabulary. Or, we could pair our Wav2Vec2 encoder with a decoder model, giving a speech encoder-decoder style model analogous to the NLP encoder-decoder model introduced in Chapter 1.
```

> A simple linear layer will give a smaller, faster overall model, but will be more susceptible to spelling and grammatical errors. Adding a decoder model greatly improves the quality of transcriptions, at the cost of a larger, slower model. This is because the decoder model is pre-trained on a large corpus of text, enabling us to leverage its learned text representations.

Reviewer: ```suggestion
A simple linear layer will give a smaller, faster overall model but will be more susceptible to spelling and grammatical errors. Adding a decoder model greatly improves the quality of transcriptions because it is pre-trained on a large corpus of text, enabling us to leverage its learned text representations. The tradeoff is a larger, slower model.
```

Since we want our system to be robust to spelling and grammar, let’s go ahead and define a speech encoder-decoder style model.

> We’ll pair a pre-trained Wav2Vec2 encoder with a pre-trained BART decoder, yielding a Wav2Vev2-2-BART model:

Reviewer: ```suggestion
We’ll pair a pre-trained Wav2Vec2 encoder with a pre-trained BART decoder, yielding a Wav2Vev2-BART model:
```

```python
from transformers import SpeechEncoderDecoderModel

encoder_id = "facebook/wav2vec2-large-lv60"
decoder_id = "facebook/bart-large"

model = SpeechEncoderDecoderModel.from_encoder_decoder_pretrained(
    encoder_id, decoder_id, encoder_add_adapter=True
)

# set special tokens for generation
model.config.decoder_start_token_id = model.decoder.config.bos_token_id
model.config.pad_token_id = model.decoder.config.pad_token_id
model.config.eos_token_id = model.decoder.config.eos_token_id
```

> The `add_adapter` argument introduces a small convolutional network between the encoder and decoder models. This adapter network helps interface the encoder and decoder by down-sampling the encoder hidden-states to better match the timescale of the decoder. In practice, including this adapter results in superior performance than the encoder and decoder models alone:

Reviewer: ```suggestion
The `encoder_add_adapter` argument introduces a small convolutional network between the encoder and decoder models. This adapter network helps interface the encoder and decoder by downsampling the encoder's hidden-states to match the timescale of the decoder better. In practice, including this adapter results in superior performance than the encoder and decoder models alone:
```

diagram

#### Preprocessing the data
Great! Now that we've defined our model we can start preparing our data for training. For this, we'll need to define two objects: a feature extractor and a tokenizer.

```python
from transformers import AutoFeatureExtractor, AutoTokenizer, Wav2Vec2Processor

# load feature extractor (for audio inputs) and tokenizer (for text outputs)
feature_extractor = AutoFeatureExtractor.from_pretrained(encoder_id)
tokenizer = AutoTokenizer.from_pretrained(decoder_id)
# combine under one class
processor = Wav2Vec2Processor(feature_extractor, tokenizer)
```

> We're already pretty familiar with the tokenizer: it converts the text data into a set of token IDs that can be interpreted by the model. This we’ll use to prepare our target labels. But the feature extractor is new! Simply put, the feature extractor prepares our input audio data.

Reviewer: Would it be more accurate to say input ids instead of token ids?

```suggestion
We're already pretty familiar with the tokenizer: it converts the text data into a set of input IDs that can be interpreted by the model. We’ll use these to prepare our target labels. But the feature extractor is new! Simply put, the feature extractor prepares our input audio data.
```

A defining feature of Wav2Vec2 is that it accepts a float array corresponding to the raw waveform of the speech signal as an input. We mentioned that the audio data is represented as a 1-dimensional array, so it's already in the right format to be read by the model (a set of continuous inputs at discrete time steps).

So, what exactly does the feature extractor do?

Well, the audio data is in the right format, but we've imposed no restrictions on the values it can take. For our model to work optimally, we want to keep all the inputs within the same dynamic range. This is going to make sure we get a similar range of activations and gradients for our samples, helping with stability and convergence during training.

> To do this, we _normalise_ our audio data, by rescaling each sample to zero mean and unit variance, a process called _feature scaling_. It's exactly this feature normalisation that our feature extractor performs!

Reviewer: ```suggestion
To do this, we _normalise_ our audio data by rescaling each sample to zero mean and unit variance, a process called _feature scaling_. It's exactly this feature normalisation that our feature extractor performs!
```

Note how we've passed the sampling rate of our audio data to our feature extractor. This is good practice, as the feature extractor performs a check under-the-hood to make sure the sampling rate of our audio data matches the sampling rate expected by the model.

Let's combine our feature extractor and tokenizer into one function to jointly preprocess our audio and text data:

```python
def prepare_dataset(batch):
    # process audio
    sample = batch["audio"]
    inputs = feature_extractor(sample["array"], sampling_rate=sample["sampling_rate"])
    # process audio length
    batch["input_values"] = inputs.input_values[0]
    batch["input_length"] = len(batch["input_values"])

    # process targets
    input_str = batch["text"].lower()
    # tokenize
    batch["labels"] = tokenizer(input_str).input_ids
    return batch

vectorized_datasets = raw_datasets.map(
    prepare_dataset,
    remove_columns=next(iter(raw_datasets.values())).column_names,
    desc="preprocess dataset",
)
```

Note that it is important to align the decoder's vocabulary with the speech transcriptions of the dataset.

> E.g. Librispeech: only captilised letters in the transcriptions, whereas BART was pretrained mostly on lower-cased text.

Reviewer: ```suggestion
For example, Librispeech only has capitalised letters in the transcriptions, whereas BART was pretrained mostly on lower-cased text.
```

> Filter to max len 20s

Reviewer: ```suggestion
Filter the audio samples to a maximum length of 20s:
```

```python
MAX_INPUT_LENGTH_IN_SECONDS = 20

max_input_length = MAX_INPUT_LENGTH_IN_SECONDS * feature_extractor.sampling_rate
# filter data that is longer than max_input_length
def is_audio_in_length_range(length):
    return length < max_input_length

vectorized_datasets = vectorized_datasets.filter(
    is_audio_in_length_range,
    input_columns=["input_length"],
)
```

> Define the datacollator:

Reviewer: ```suggestion
Define the data collator:
```
